### PR TITLE
Add support for Microsoft extensions '1.3.6.1.4.1.311.20.2', '1.3.6.1…  ….4.1.311.21.1' and '1.3.6.1.4.1.311.21.2'

### PR DESCRIPTION
**Pull request by MauAraujo**
I added support for the extensions so that they match the structure in ad1b88e. I tested them and they all display properly as shown: 
Captura de pantalla 2019-03-25 a la(s) 13 08 38

### CODE REVIEW

File: src/viewer/js/der.js [hunk 1]
```diff
@@ -355,11 +361,35 @@ export const parse = async (der) => {
...
+  }
+
+  if(msCrypto.certServerPreviousHash) {
+    msCrypto.certServerPreviousHash = {
+      critical: criticalExtensions.includes('1.3.6.1.4.1.311.21.2'),
+      hash: hashify(msCrypto.certServerPreviousHash.valueBlock.valueHex),
+    };
+  }
+
+  msCrypto.exists = (msCrypto.certificatePolicies || msCrypto.certificateTemplate
```
Comment (april) [inline src/viewer/js/der.js]: I would probably place each of these across their own lines at this point.

File: src/viewer/js/der.js [hunk 2]
```diff
@@ -355,11 +361,35 @@ export const parse = async (der) => {
       critical: criticalExtensions.includes('1.3.6.1.4.1.311.21.7'),
       id: msCrypto.certificateTemplate.extnID,
       major: msCrypto.certificateTemplate.templateMajorVersion,
-      minor: msCrypto.certificateTemplate.templateMinorVersion,     
+      minor: msCrypto.certificateTemplate.templateMinorVersion,
     };
   }
 
-  msCrypto.exists = (msCrypto.certificatePolicies || msCrypto.certificateTemplate) ? true : false;
+  if(msCrypto.enrollmentInfrastructure) {
```
Comment (april) [inline src/viewer/js/der.js]: Please keep the spacing consistent with the rest of the code.

File: src/viewer/js/der.js [hunk 3]
```diff
@@ -60,6 +60,9 @@ export const parse = async (der) => {
   const supportedExtensions = [
     '1.3.6.1.4.1.311.21.7',     // microsoft certificate template
     '1.3.6.1.4.1.311.21.10',    // microsoft certificate policies
+    '1.3.6.1.4.1.311.20.2',     // microsoft enrollment infrastructure certificate type
```
Comment (april) [inline src/viewer/js/der.js]: These are in numerical order for legibility.  :)

File: src/viewer/js/der.js [hunk 4]
```diff
@@ -60,6 +60,9 @@ export const parse = async (der) => {
...
+    '1.3.6.1.4.1.311.21.1',     // microsoft certificate server ca version
+    '1.3.6.1.4.1.311.21.2',      // microsoft certificate server previous hash
```
Comment (april) [inline src/viewer/js/der.js]: Please keep the same spacing.

File: src/viewer/js/der.js [hunk 5]
```diff
@@ -341,6 +344,9 @@ export const parse = async (der) => {
   let msCrypto = {
     certificatePolicies: getX509Ext(x509.extensions, '1.3.6.1.4.1.311.21.10').parsedValue,
     certificateTemplate: getX509Ext(x509.extensions, '1.3.6.1.4.1.311.21.7').parsedValue,
+    enrollmentInfrastructure: getX509Ext(x509.extensions, '1.3.6.1.4.1.311.20.2').parsedValue,
```
Comment (april) [inline src/viewer/js/der.js]: These should be in some kind of order, either numerical by OID or alphabetical by key.

File: src/viewer/index.handlebars [hunk 6]
```diff
@@ -410,6 +410,51 @@
...
+            {{#if (truthy this.ext.msCrypto.enrollmentInfrastructure)}}
+              <div class="panel-section-subheader">
+                <span>
+                  {{#if this.ext.msCrypto.enrollmentInfrastructure.critical}}<img class="critical" alt="critical extension" title="This extension has been marked as critical, meaning that clients must reject the certificate if they don't understand it." src="../icons/critical.svg">{{/if~}}
+                  Enrollment Infrastructure
+                </span>
+              </div>
+              <div class="panel-list-item">
+                <div class="text">Certificate Type</div>
+                <div class="text">{{this.ext.msCrypto.enrollmentInfrastructure.certType}}</div>
```
Comment (april) [inline src/viewer/index.handlebars]: Please keep the same spacing as the rest of the code.

File: src/viewer/index.handlebars [hunk 7]
```diff
@@ -410,6 +410,51 @@
...
+                <div class="text">{{this.ext.msCrypto.certServerCAVersion.keyIndex}}</div>
+              </div>
+            {{/if}}
+
+            {{! Certificate Server Previous Hash }}
+            {{#if (truthy this.ext.msCrypto.certServerPreviousHash)}}
+              <div class="panel-section-subheader">
+                <span>
+                  {{#if this.ext.msCrypto.certServerPreviousHash.critical}}<img class="critical" alt="critical extension" title="This extension has been marked as critical, meaning that clients must reject the certificate if they don't understand it." src="../icons/critical.svg">{{/if~}}
+                  Certificate Server Previous Version
```
Comment (april) [inline src/viewer/index.handlebars]: When you look at the rest of the entries, you'll notice that they don't split across lines for aesthetic purposes.  :)

File: src/viewer/js/der.js [hunk 8]
```diff
@@ -355,11 +361,35 @@ export const parse = async (der) => {
...
-      minor: msCrypto.certificateTemplate.templateMinorVersion,     
+      minor: msCrypto.certificateTemplate.templateMinorVersion,
     };
   }
 
-  msCrypto.exists = (msCrypto.certificatePolicies || msCrypto.certificateTemplate) ? true : false;
+  if(msCrypto.enrollmentInfrastructure) {
+    msCrypto.enrollmentInfrastructure = {
+      critical: criticalExtensions.includes('1.3.6.1.4.1.311.20.2'),
+      certType: msCrypto.enrollmentInfrastructure.valueBlock.value,
```
Comment (april) [inline src/viewer/js/der.js]: We probably shouldn't use abbreviations like "subCA", we should write out what it means.

**Comment by MauAraujo** [comment]
Alright, I have fixed the spacing issues and will start to work on displaying the full certificate types in the enrollment extension.

**Comment by MauAraujo** [comment]
I used this document: 
https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-CRTD/%5bMS-CRTD%5d.pdf
that I downloaded from this microsoft page: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-crtd/4c6950e4-1dc2-4ae3-98c3-b8919bb73822
to display the full name of the certificate type instead of an abbreviation. This change is shown here: 
certType
It now displays "Subordinate Certification Authority" instead of just "SubCA".

**Pull request closed without merging**